yksgi/src/lib.rs

```diff
@@ -313,8 +316,68 @@ impl SGInterp {
             // FIXME: When we see the control point we are done and can just return.
             return true;
         } else {
-            // FIXME: Properly implement calls.
-            todo!()
+            // We are going to do a native call via libffi, starting with looking up the address of
+            // the callee in the virtual address space.
+            //
+            // OPT: This could benefit from caching, and we may already know the address of the
+            // callee if we inlined it when preparing traces:
+            // https://github.com/ykjit/yk/issues/544
+            assert!(!LLVMIsAFunction(func).is_null());
+            let fptr = dlsym(ptr::null_mut(), name.as_ptr() as *const i8);
+            if fptr == ptr::null_mut() {
+                todo!("couldn't find symbol: {}", name);
+            }
+
+            if LLVMIsFunctionVarArg(LLVMGetElementType(LLVMTypeOf(func))) != 0 {
+                todo!("calling a varargs function");
+            }
+
+            // Now build the calling interface and collect the values of the callee's arguments.
+            //
+            // It may be tempting to use `LLVMGetArgOperand()` to extract callee operands from our
+            // `CallInst`, but at the time of writing this wraps the C++ method
+            // `FuncletPadInst::getArgOperand()` which is not what we want.
+            //
+            // `CallBase::getArgOperand()` isn't currently exposed by the LLVM C API, so we instead
+            // use the implicit knowledge that for a `CallInst` the first `N` operands are the
+            // operands of the function being called (the remaining operand is the callee).
+            //
+            // Despite this, `LLVMGetNumArgOperands()` *is* generic and does the right thing for
+            // any callable function-like thing, so we can still use that for the loop bounds.
+            let mut builder = FFIBuilder::new();
+            let mut arg_vals = Vec::new();
+            for i in 0..LLVMGetNumArgOperands(call_inst) {
+                let arg = Value::new(LLVMGetOperand(call_inst, i.try_into().unwrap()));
+                builder = builder.arg(arg.get_ffi_type());
+                // Note that a `FFIArg()` is unsafe in that caches a raw pointer to something which
+                // may later die. Converting to `FFIArg()` in this loop would cause UB later
+                // because `&self.var_lookup(&arg).val as *const _` would be a dangling pointer
+                // once the loop body ends.
+                arg_vals.push(self.var_lookup(&arg).val);
+            }
+            let ret_val = Value::new(call_inst);
+            builder = builder.res(ret_val.get_ffi_type());
+            let cif = builder.into_cif(); // OPT: cache CIFs for repeated calls to same func sig.
+
+            // Actually do the call.
+            let ret_ty = ret_val.get_type();
+            let ffi_arg_vals: Vec<FFIArg> = arg_vals.iter().map(|a| ffi_arg(a)).collect();
```

Comment:
An `into_iter` here *might* help the compiler reuse the underlying Vec? I might be too hopeful though!

Reply:
Wouldn't we risk destructing the old vector, leaving not one dangling pointer, but a whole vector of them?

Reply:
If it's not safe, Rust will tell you.

Reply:
Wrong. Creating a dangling pointer is safe in Rust. If we then pass that dangling pointer into an unsafe interface like `CIF::call()` (as we do here) then boom.

Reply:
Hang on, what will dangle? `Vec.into_iter` is safe Rust: nothing is left dangling.

Reply:
`ffi_arg` caches the reference argument you pass it as a raw pointer, so if you consume what is referenced, then the pointer dangles.

See:
 - [here](https://docs.rs/libffi/latest/src/libffi/middle/mod.rs.html#49-51)
 - [and here](https://docs.rs/libffi/latest/src/libffi/middle/mod.rs.html#40-42)

Reply:
(Note that I imported `arg()` as `ffi_arg()` for clarity)

Reply:
> ffi_arg() caches the reference argument you pass it as a raw pointer

Then this code is undefined behaviour: there is no guarantee that a `Vec`s backing memory stays at a stable address.

Reply:
I don't see how `libffi` crate could work if that were the case.

Is it not the case that a `Vec`'s backing storage stays put if you don't mutate the vector?

Reply:
> Is it not the case that a Vec's backing storage stays put if you don't mutate the vector?

It kind of depends how you read https://doc.rust-lang.org/std/vec/struct.Vec.html -- I would be pessimistic and say "no" as it only says that *some* operations won't change the backing store. That's a weaker guarantee than I suspect was intended, but it is what it is.

Personally, I think that if you want to refer to the *address* of something, you must be in control of the allocation (e.g. a `Box` is fixed in memory), otherwise you're at the mercy of other code.

Reply:
That would be horribly inefficient. There must be a better way. This is the whole point of libffi -- you don't know the shape of the function until runtime.

The [example in the `libffi` docs](https://docs.rs/libffi/latest/libffi/middle/struct.Cif.html#examples) uses an array. What if we make a slice from the vec so that the compiler can't possibly move the backing storage?

Reply:
Let's try and split two things apart.

First, `ffi::arg` is this horror https://docs.rs/libffi/latest/src/libffi/middle/mod.rs.html#49 -- IMHO that should be an `unsafe` function because converting `&T` to `*const T` does not communicate to the user that they can't move `&T` afterwards. [It would be better, probably, if `Arg::new` took in a `*const T` in the first place.]

Second, the example in https://docs.rs/libffi/latest/libffi/middle/struct.Cif.html#examples a) does not create a `Vec`, it creates a fixed-size array b) does not reference addresses in that fixed-size array.

The problem is that we're effectively doing this:

```rust
let v = vec![1,2,3];
let addr = &v[0] as *const _;
```

If the `Vec` moves its backing store -- and the docs do not say that it won't -- then `addr` can point to an invalid address and we're in undefined behaviour. This is the situation we're in.

Just to reiterate: the interace to `libffi::Arg::new` seems designed to let seemingly safe Rust code descend into actually unsafe Rust code without making clear that's happened. Eek! Also, allocating a chunk of memory for the args will be *exactly as efficient as creating a Vec to store them* (we know how big the chunk will be at allocation time), so there's no real reason not to do it.

Reply:
After a lot of offline back and forth, we've decided to use a manually allocated chunk of memory (which can't move).

Hopefully dc39229b does the trick. Tests pass, but I'm unsure if getting a mutable reference to the slice and writing into it is correct. Please @ltratt can you have a look?